just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java

```diff
@@ -576,24 +580,35 @@ private MediaSource getAudioSource(final Object json) {
     }
 
     private MediaSource decodeAudioSource(final Object json) {
-        Map<?, ?> map = (Map<?, ?>)json;
-        String id = (String)map.get("id");
+        Map<?, ?> map = (Map<?, ?>) json;
+        String id = (String) map.get("id");
+
+        DrmSessionManager drmSessionManager = null;
+        if (map.get("drm") != null) {
+            Map<?, ?> drm = (Map<?, ?>) map.get("drm");
+            if (drm.get("clearKey") != null) {
+                drmSessionManager = new DefaultDrmSessionManager.Builder()
+                        .setUuidAndExoMediaDrmProvider(C.CLEARKEY_UUID, FrameworkMediaDrm.DEFAULT_PROVIDER).
+                                build(new LocalMediaDrmCallback(((String) drm.get("clearKey")).getBytes()));
+            }
```

Comment:
How would this design be extended to support widevine with callbacks?

Reply:
The idea would be:

For widewine a token is used for authentication and needs to be set when creating the DefaultDrmSessionManager.

In the Dart code when constructing the drm map it would look something like.

{"widewine-token","myToken"}//others to be aded.  A method could be added to DRMHelper class widewine(String token) which would return the map.

In AudioPlayer.java

After the check if DRM map is set sequentially check for properties(e.g. widewine-token and clearKey("as is done currently)

Depending on sequence of the checks the correct drmSessionManager would be set which is used when construction the MediaSource from the factory.


Hope this makes sense?

Reply:
But how would an app implement `MediaDrmCallback` if it had special configuration requirements?

Reply:
Actually, not sure if that callback is used anymore, but there is still `DrmSessionEventListener.EventDispatcher`.

Reply:
I have not seen that the MediaDrmCallBack and EventDispatcher is used during my research on this. Most of the DRM is handled internally by ExoPlayer the configuration passed. 

Below and example how widewine can be implemented:(taken from better_player). I have added the ClearKey DRM to this project.



HttpMediaDrmCallback httpMediaDrmCallback =
                    new HttpMediaDrmCallback(licenseUrl, new DefaultHttpDataSource.Factory());

            if (drmHeaders != null) {
                for (Map.Entry<String, String> entry : drmHeaders.entrySet()) {
                    httpMediaDrmCallback.setKeyRequestProperty(entry.getKey(), entry.getValue());
                }
            }

                UUID drmSchemeUuid = Util.getDrmUuid("widevine");
                if (drmSchemeUuid != null) {
                    drmSessionManager =
                            new DefaultDrmSessionManager.Builder()
                                    .setUuidAndExoMediaDrmProvider(drmSchemeUuid,
                                            uuid -> {
                                                try {
                                                    FrameworkMediaDrm mediaDrm = FrameworkMediaDrm.newInstance(uuid);
                                                    // Force L3.
                                                    mediaDrm.setPropertyString("securityLevel", "L3");
                                                    return mediaDrm;
                                                } catch (UnsupportedDrmException e) {
                                                    return new DummyExoMediaDrm();
                                                }
                                            })
                                    .setMultiSession(false)
                                    .build(httpMediaDrmCallback);